### PR TITLE
Contextualize type splices

### PR DESCRIPTION
**Pull request by nicolasstucki**
Change this line https://github.com/lampepfl/dotty/pull/6958/files#diff-78442038b1a8047647ef95723ef2a21bR22 and handle it in the compiler

### CODE REVIEW

File: library/src/scala/runtime/quoted/Unpickler.scala [hunk 1]
```diff
@@ -19,5 +19,6 @@ object Unpickler {
   /** Unpickle `repr` which represents a pickled `Type` tree,
    *  replacing splice nodes with `args`
    */
-  def unpickleType[T](repr: Pickled, args: Seq[Seq[Any] => Type[_]]): Type[T] = new TastyType[T](repr, args)
+  def unpickleType[T](repr: Pickled, args: Seq[Seq[Any] => Type[_]]): given QuoteContext => Type[T] = new TastyType[T](repr, args)
```
Comment (liufengyun) [inline library/src/scala/runtime/quoted/Unpickler.scala]: Just curious, what's the recommended practice to use `given`: as parameters or contextual function types. Semantically the two are equivalent, and performance-wise they should be almost the same due to the optimization for implicit functions via direct methods.

File: tests/pos/i4414.scala [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import scala.quoted._
 
 object Test {
+  given as QuoteContext = ???
```
Comment (liufengyun) [inline tests/pos/i4414.scala]: It's not obvious to me why this is needed, given that all implicits are available and type check, and there are no quotes nor splices.

Comment (nicolasstucki) [reply tests/pos/i4414.scala]: The calls to `a` and `b` synesize some quoted types `[...]`, those require a context.

File: library/src/scala/runtime/quoted/Unpickler.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/runtime/quoted/Unpickler.scala]: Here the desison was taken to have the symmetry between the quote and splices. The splice takes a contextual function, therefore the quote returns a contextual function. In this particular case, the alignment also helped to implement the quo/splice cancellation.

**Pull request merged**